### PR TITLE
use update_ff and <<= for sequential blocks

### PR DESCRIPTION
**Pull request by jsn1993**
We will shift away from the old way of doing pure constrained based scheduling for sequential logic, as Python's value assignment gave us a lot of trouble and we had to use deepcopies to (almost) guarantee correct execution.

### CODE REVIEW

File: pymtl3/datatypes/BitStruct.py [hunk 1]
```diff
@@ -94,6 +81,13 @@ class {name}( BitStruct ):
   def __init__( s, {args_str} ):
 {assign_str}
 
+  def __ilshift__( s, x ):
```
Comment (jsn1993) [inline pymtl3/datatypes/BitStruct.py]: @yo96 Can you also implement this recursive ilshift and recursive _flip in the new struct branch?

File: pymtl3/datatypes/bitstructs.py [hunk 2]
```diff
@@ -269,6 +269,58 @@ def _mk_hash_fn( fields ):
...
+    if isinstance( type_, list ):
+      i = 0
+      loop = f"{' '*i}for i{i} in range({len(type_)}):"
+      ilshift_strs.append(loop)
+      flip_strs   .append(loop)
+      type_ = type_[0]
+      i = 1
+      while isinstance( type_, list ):
+        loop = f"{' '*(i*2)}for i{i} in range({len(type_)}):"
+        ilshift_strs.append(loop)
```
Comment (jsn1993) [inline pymtl3/datatypes/bitstructs.py]: @yo96 i guess you should review this bitstruct file

File: .travis.yml [hunk 3]
```diff
@@ -52,11 +53,12 @@ script:
  - git diff --exit-code
  - mkdir -p build
  - cd build
- - py.test --cov-report xml --cov=pymtl3 ../pymtl3 ../examples $CONFIG --hypothesis-profile CI
+ - export NOT_USING_PYPY=$(python -V | grep -q "PyPy"; echo $?)
```
Comment (jsn1993) [inline .travis.yml]: @cbatten @ptpan This is how I turn off coverage for pypy. Looks kinda hacky. Feedbacks are welcome.

File: pymtl3/datatypes/bits_import.py [hunk 4]
```diff
@@ -26,49 +24,43 @@
...
+              filename="bits_import.py", mode="exec") )
 
 def mk_bits( nbits ):
-  assert nbits < 16384, "We don't allow bitwidth to exceed 16384."
+  assert nbits < 512, "We don't allow bitwidth to exceed 512."
   if nbits in _bits_types:  return _bits_types[ nbits ]
-
-  exec((py.code.Source( bits_template.format( **vars() ) ).compile()), globals())
-  return _bits_types[ nbits ]
+  _locals = {}
```
Comment (ptpan) [inline pymtl3/datatypes/bits_import.py]: Remove unnecessary line

File: pymtl3/datatypes/bits_import.py [hunk 5]
```diff
@@ -26,49 +24,43 @@
...
+  exec(compile( bits_template.format(nbits), filename="Generated Bits", mode="exec" ))
```
Comment (ptpan) [inline pymtl3/datatypes/bits_import.py]: Generated BitsN?

File: pymtl3/datatypes/bitstructs.py [hunk 6]
```diff
@@ -267,6 +267,58 @@ def _mk_hash_fn( fields ):
     [ f'return hash({self_tuple})' ]
   )
 
+#-------------------------------------------------------------------------
+# _mk_ff_fn
+#-------------------------------------------------------------------------
+#
+# def __ilshift__( self, other ):
```
Comment (ptpan) [inline pymtl3/datatypes/bitstructs.py]: Mark this as an example

File: pymtl3/datatypes/bitstructs.py [hunk 7]
```diff
@@ -412,6 +464,18 @@ def _convert_list_to_tuple( x ):
...
+  assert not '__ilshift__' in cls.__dict__ and not '_flip' in cls.__dict__
+
+  cls.__ilshift__, cls._flip = _mk_ff_fn( fields )
+
+  def get_field_type( cls, name ):
+    if name in cls.__bitstruct_fields__:
+      return cls.__bitstruct_fields__[ name ]
+    raise AttributeError( f"{cls} has no field '{name}'" )
+
+  cls.get_field_type = classmethod(get_field_type)
```
Comment (ptpan) [inline pymtl3/datatypes/bitstructs.py]: For consistency you probably need to assert `get_field_type` is not in the class attributes

File: pymtl3/dsl/AstHelper.py [hunk 8]
```diff
@@ -124,13 +130,40 @@ def _get_full_name( self, node ):
...
+  def visit_Assign( self, node ):
+
+    if self.is_update_ff:
+      assert len( node.targets ) == 1
+      target = node.targets[0]
+
+      if isinstance( target, (ast.Attribute, ast.Subscript) ):
+        while isinstance( target, (ast.Attribute, ast.Subscript) ):
+          target = target.value
+        assert isinstance( target, ast.Name ), "Please call pymtl3 developers"
```
Comment (ptpan) [inline pymtl3/dsl/AstHelper.py]: Improve this (a test case?)

File: pymtl3/dsl/Component.py [hunk 9]
```diff
@@ -468,10 +468,10 @@ def lock_in_simulation( s ):
           if isinstance( obj, Signal ):
             try:
               current_obj[i] = obj.default_value()
-            except Exception as err:
-              err.message = repr(obj) + " -- " + err.message
-              err.args = (err.message,)
-              raise err
+              current_obj[i] <<= obj.default_value()
+            except Exception as e:
+              print("For object", repr(obj))
```
Comment (ptpan) [inline pymtl3/dsl/Component.py]: Manipulate the message inside `e`?

File: pymtl3/dsl/ComponentLevel2.py [hunk 10]
```diff
@@ -165,13 +168,27 @@ def lookup_variable( obj, name_depth, node_depth ):
...
+                raise InvalidFFAssignError( s, func, nodelist[0].lineno,
+                  "has a wrong type on the left value. "
+                  "We only allow <<= on Bits/BitStruct type signals, not {x._dsl.Type}")
 
         # This is a function call without "s." prefix, check func list
         elif obj_name[0][0] in s._dsl.name_func:
           call = s._dsl.name_func[ obj_name[0][0] ]
           all_objs.add( call )
-          s._dsl.astnode_objs[ nodelist[0] ].append( call )
+          # s._dsl.astnode_objs[ nodelist[0] ].append( call )
```
Comment (ptpan) [inline pymtl3/dsl/ComponentLevel2.py]: probably should remove this comment since it's out-dated

File: pymtl3/dsl/ComponentLevel2.py [hunk 11]
```diff
@@ -187,7 +204,7 @@ def lookup_variable( obj, name_depth, node_depth ):
 
     # what object each astnode corresponds to. You can't have two update
     # blocks in one component that have the same ast.
-    s._dsl.astnode_objs = defaultdict(list)
+    # s._dsl.astnode_objs = defaultdict(list)
```
Comment (ptpan) [inline pymtl3/dsl/ComponentLevel2.py]: you might want to take a pass to remove these comments related to astnode_obj

File: pymtl3/dsl/NamedObject.py [hunk 12]
```diff
@@ -143,17 +143,68 @@ def __setattr_for_elaborate__( s, name, obj ):
...
 
     if name[0] != '_': # filter private variables
-      stack = deque( [ (obj, []) ] )
-      while stack:
-        u, indices = stack.popleft()
+      sd = s._dsl
 
-        if isinstance( u, NamedObject ):
-            sd = s._dsl
+      if isinstance( obj, NamedObject ):
```
Comment (ptpan) [inline pymtl3/dsl/NamedObject.py]: Maybe a comment here to explain this optimization

File: pymtl3/dsl/test/ComponentAPI_test.py [hunk 13]
```diff
@@ -320,3 +320,17 @@ def line_trace( s ):
   a.in_ = Bits32(10)
   a.tick()
   assert a.out == 10 + 444 * 2
+
+def test_garbage_collection():
```
Comment (ptpan) [inline pymtl3/dsl/test/ComponentAPI_test.py]: Disable this test at least in CI?

File: pymtl3/passes/rtlir/behavioral/BehavioralRTLIRGenL1Pass.py [hunk 14]
```diff
@@ -46,7 +46,7 @@ def __call__( s, m ):
 class BehavioralRTLIRGeneratorL1( ast.NodeVisitor ):
   def __init__( s, component ):
     s.component = component
-    s.mapping   = component.get_astnode_obj_mapping()
+    # s.mapping   = component.get_astnode_obj_mapping()
```
Comment (ptpan) [inline pymtl3/passes/rtlir/behavioral/BehavioralRTLIRGenL1Pass.py]: Remove comments related to astnode_obj?

**Pull request merged**